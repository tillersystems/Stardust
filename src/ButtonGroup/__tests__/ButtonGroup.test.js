import React from 'react';
import { fireEvent } from '@testing-library/react';

import { Button, ButtonGroup } from '../..';

const originalWarn = console.error;
afterEach(() => (console.error = originalWarn));

describe('<ButtonGroup />', () => {
  let consoleOutput = [];
  const mockedWarn = output => consoleOutput.push(output);
  beforeEach(() => (console.error = mockedWarn));

  test('should render without a problem', () => {
    const { container } = render(
      <ButtonGroup>
        <Button name="ON" appearance="secondary">
          ON
        </Button>
        <Button name="OFF" appearance="secondary">
          OFF
        </Button>
      </ButtonGroup>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render with a default value', () => {
    const { getByText } = render(
      <ButtonGroup defaultActiveButtonName="ON">
        <Button name="ON" appearance="secondary">
          ON
        </Button>
        <Button name="OFF" appearance="secondary">
          OFF
        </Button>
      </ButtonGroup>,
    );
    const offButton = getByText('ON');

    expect(offButton).toHaveAttribute('data-isactive', 'true');
  });

  test('should not clone child if disabled', () => {
    const { getByText } = render(
      <ButtonGroup defaultActiveButton="ON">
        <Button name="ON" appearance="secondary">
          ON
        </Button>
        <Button name="OFF" appearance="secondary" disabled>
          OFF
        </Button>
      </ButtonGroup>,
    );

    const offButton = getByText('OFF');

    expect(offButton).toHaveAttribute('disabled');
    expect(offButton).not.toHaveAttribute('data-isactive', 'true');
    expect(offButton).not.toHaveAttribute('data-appearance', 'secondary');
  });

  test('should not clone child if not a button element', () => {
    const { getByText } = render(
      <ButtonGroup defaultActiveButton="ON">
        <div name="ON">ON</div>
        <div name="OFF">OFF</div>
      </ButtonGroup>,
    );

    const onButton = getByText('ON');
    const offButton = getByText('OFF');

    expect(onButton).not.toHaveAttribute('data-isactive', 'true');
    expect(onButton).not.toHaveAttribute('data-appearance', 'secondary');
    expect(offButton).not.toHaveAttribute('data-isactive', 'true');
    expect(offButton).not.toHaveAttribute('data-appearance', 'secondary');
  });

  test('should respond to a click handler', () => {
    const spy = jest.fn();
    const { getByText } = render(
      <ButtonGroup defaultActiveButton="ON" onChange={spy}>
        <Button name="ON" appearance="secondary">
          ON
        </Button>
        <Button name="OFF" appearance="secondary">
          OFF
        </Button>
      </ButtonGroup>,
    );

    const offButton = getByText('OFF');

    fireEvent.click(offButton);

    expect(spy).toHaveBeenCalledTimes(1);
  });

  test('should call the onclick handler on child button', () => {
    const spy = jest.fn();
    const spy2 = jest.fn();
    const { getByText } = render(
      <ButtonGroup defaultActiveButton="ON" onChange={spy}>
        <Button name="ON" appearance="secondary">
          ON
        </Button>
        <Button name="OFF" appearance="secondary" onClick={spy2}>
          OFF
        </Button>
      </ButtonGroup>,
    );

    const offButton = getByText('OFF');

    fireEvent.click(offButton);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledTimes(1);
  });

  test('should throw an error if child has not name props', () => {
    const renderComponent = () =>
      render(
        <ButtonGroup defaultActiveButtonName="ON">
          <Button appearance="secondary">ON</Button>
          <Button appearance="secondary">OFF</Button>
        </ButtonGroup>,
      );
    expect(renderComponent).toThrow('"name" prop must be provided to Button');
  });
});
