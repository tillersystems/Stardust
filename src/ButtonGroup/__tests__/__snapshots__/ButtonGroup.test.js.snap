// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> should render without a problem 1`] = `
.c2 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.75rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  color: hsl(207,13%,45%);
  background: linear-gradient(180deg,hsl(0,100%,100%) 0%,hsl(216,45%,98%) 100%);
  border: 1px solid hsl(207,22%,90%);
  padding: 0.75rem 1.6rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .c1 {
  border-radius: 0;
}

.c0 .c1:first-of-type {
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
}

.c0 .c1:nth-child(n):not(:first-of-type) {
  border-left: 0;
}

.c0 .c1:last-of-type {
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
  border-left: 0;
}

.c0 .c1[data-isactive='true'] {
  text-shadow: 0.6px 0 0 currentColor;
}

.c0 .c1[data-isactive='true'][data-appearance='secondary'] {
  color: hsl(200,74%,46%);
}

.c0 .c1[data-isactive='true'][data-appearance='secondary'] svg > path {
  fill: rgba(30,146,204,0.7);
}

.c0 .c1[data-isactive='false'] {
  color: rgba(255,255,255,0.7);
}

.c0 .c1[data-isactive='false'] svg > path {
  fill: rgba(255,255,255,0.7);
}

.c0 .c1[data-isactive='false'][data-appearance='secondary'],
.c0 .c1[data-isactive='false'][data-appearance='default'] {
  color: rgba(100,116,130,0.9);
}

.c0 .c1[data-isactive='false'][data-appearance='secondary'] svg > path,
.c0 .c1[data-isactive='false'][data-appearance='default'] svg > path {
  fill: rgba(100,116,130,0.7);
}

<div
  class="c0"
  role="group"
  tabindex="-1"
>
  <button
    class="c1 c2"
    data-appearance="secondary"
    data-isactive="false"
    name="ON"
    type="button"
  >
    ON
  </button>
  <button
    class="c1 c2"
    data-appearance="secondary"
    data-isactive="false"
    name="OFF"
    type="button"
  >
    OFF
  </button>
</div>
`;
